src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableList_1.cs

```diff
@@ -3226,48 +3189,135 @@ private static Node BalanceNode(Node node)
                 }
                 else
                 {
-                    if (left != null)
-                    {
-                        _left = left;
-                    }
+                    _left = left;
+                    _right = right;
+                    _height = ParentHeight(left, right);
+                    _count = ParentCount(left, right);
+                    return this;
+                }
+            }
 
-                    if (right != null)
-                    {
-                        _right = right;
-                    }
+            /// <summary>
+            /// Creates a node mutation, either by mutating this node (if not yet frozen) or by creating a clone of this node
+            /// with the described changes.
+            /// </summary>
+            /// <param name="left">The left branch of the mutated node.</param>
+            /// <returns>The mutated (or created) node.</returns>
+            private Node MutateLeft(Node left)
+            {
+                Requires.NotNull(left, nameof(left));
+                Debug.Assert(!this.IsEmpty);
 
-                    _height = checked((byte)(1 + Math.Max(_left._height, _right._height)));
-                    _count = 1 + _left._count + _right._count;
+                if (_frozen)
+                {
+                    return new Node(_key, left, _right);
+                }
+                else
+                {
+                    _left = left;
+                    _height = ParentHeight(left, _right);
+                    _count = ParentCount(left, _right);
+                    return this;
+                }
+            }
+
+            /// <summary>
+            /// Creates a node mutation, either by mutating this node (if not yet frozen) or by creating a clone of this node
+            /// with the described changes.
+            /// </summary>
+            /// <param name="right">The right branch of the mutated node.</param>
+            /// <returns>The mutated (or created) node.</returns>
+            private Node MutateRight(Node right)
+            {
+                Requires.NotNull(right, nameof(right));
+                Debug.Assert(!this.IsEmpty);
+
+                if (_frozen)
+                {
+                    return new Node(_key, _left, right);
+                }
+                else
+                {
+                    _right = right;
+                    _height = ParentHeight(_left, right);
+                    _count = ParentCount(_left, right);
                     return this;
                 }
             }
 
             /// <summary>
+            /// Calculates the height of the parent node from its children.
+            /// </summary>
+            /// <param name="left">The left child.</param>
+            /// <param name="right">The right child.</param>
+            /// <returns>The height of the parent node.</returns>
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
```

Comment:
I checked and this was not getting inlined, so I applied AggressiveInlining. There was a significant perf difference (~5% for `Insert` and `Add`) when I did so.